cpp/include/cuspatial/detail/multilinestring_segment.cuh

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/detail/functors.cuh>
+#include <cuspatial/detail/range/multilinestring_segment_range.cuh>
+#include <cuspatial/detail/utility/zero_data.cuh>
+#include <cuspatial/iterator_factory.cuh>
+#include <cuspatial/range/range.cuh>
+#include <cuspatial/traits.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <thrust/device_vector.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/logical.h>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @internal
+ * @brief Functor that returns true if current value is greater than 0.
+ */
+template <typename IndexType>
+struct greater_than_zero_functor {
+  __device__ IndexType operator()(IndexType x) const { return x > 0; }
+};
+
+/**
+ * @internal
+ * @brief Owning class to provide iterators to segments in a multilinestring range
+ *
+ * The owned memory in this struct is `_non_empty_linestring_prefix_sum`, which equals the
+ * number of linestrings in the multilinestring range plus 1. This vector holds the number
+ * of non empty linestrings that precedes the current linestring.
+ *
+ * This class is only meant for tracking the life time of the owned memory. To access the
+ * segment iterators, call `view()` function to create a non-owning object of this class.
+ *
+ * For detailed explanation on the implementation of the segment iterators, see documentation
+ * of `multilinestring_segment_range`.
+ *
+ * @note To use this class with a multipolygon range, cast the multipolygon range as a
+ * multilinestring range.
+ *
+ * TODO: Optimization: for range that does not contain any empty linestrings,
+ * `_non_empty_linestring_prefix_sum` can be substituted with `counting_iterator`.
+ *
+ * @tparam MultilinestringRange The multilinestring range to initialize this class with.
+ */
+template <typename MultilinestringRange>
+class multilinestring_segment {
```

Comment:
This name sounds like it is a single segment. What this class really provides is just storage for a segment range. So perhaps we name it as such?

```suggestion
class multilinestring_segment_range_storage {
```

Reply:
My favorite of those is `multilinestring_segment_manager`.  Did you actually show the code for the class to ChatGPT and it wrote these?

Reply:
No. This is my prompt:
```
What are the common naming techniques to show that a C++ class is an owning object?
```

Then it showed a few techniques. I made these names based on its advice.